### PR TITLE
Scripting API for Render Graph

### PR DESCRIPTION
**Pull request by JulienIgnace-Unity**
### Purpose of this PR
Scripting API for Render Graph

### CODE REVIEW

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs [hunk 1]
```diff
@@ -17,7 +17,7 @@ public enum DepthAccess
     }
 
     /// <summary>
-    /// Context passed to every render pass.
+    /// Use this struct to specify the context given to every render pass.
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraph.cs]: @JordanL8  Actually, the system automatically fills it and provides it to the user function for each render pass

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 2]
```diff
@@ -3,13 +3,23 @@
 
 namespace UnityEngine.Experimental.Rendering.RenderGraphModule
 {
+    /// <summary>
+    /// The RenderGraphBuilder is the class that allows users to setup a new Render Pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Use this struct to set up a new Render Pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 3]
```diff
@@ -3,13 +3,23 @@
...
+    /// </summary>
     public struct RenderGraphBuilder : IDisposable
     {
         RenderGraph.RenderPass      m_RenderPass;
         RenderGraphResourceRegistry m_Resources;
         bool                        m_Disposed;
 
         #region Public Interface
+        /// <summary>
+        /// Specify that a texture resource will be used as a color render target during the pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify that the pass should use a Texture resource as a color render target.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 4]
```diff
@@ -3,13 +3,23 @@
...
+        /// This will have the same effect as WriteTexture and will also automatically set the texture as a render target
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: This has the same effect as WriteTexture and also automatically sets the Texture to use as a render target.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 5]
```diff
@@ -3,13 +3,23 @@
...
+        /// </summary>
+        /// <param name="input">Texture resource that will be used as a color render target</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: The Texture resource to use as a color render target.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 6]
```diff
@@ -20,6 +30,12 @@ public RenderGraphMutableResource UseColorBuffer(in RenderGraphMutableResource i
             return input;
         }
 
+        /// <summary>
+        /// Specify that a texture resource will be used a depth buffer during the pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify that the pass should use a Texture resource as the depth buffer.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 7]
```diff
@@ -20,6 +30,12 @@ public RenderGraphMutableResource UseColorBuffer(in RenderGraphMutableResource i
...
+        /// </summary>
+        /// <param name="input">Texture resource that will be used as a depth buffer</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: The Texture resource to use as a depth buffer.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 8]
```diff
@@ -20,6 +30,12 @@ public RenderGraphMutableResource UseColorBuffer(in RenderGraphMutableResource i
...
+        /// <param name="flags">Specify if the depth buffer will be read, written to, or both</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify the access level for the depth buffer. This allows you to say whether you can read from or write to the depth buffer, or do both.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 9]
```diff
@@ -33,6 +49,11 @@ public RenderGraphMutableResource UseDepthBuffer(in RenderGraphMutableResource i
             return input;
         }
 
+        /// <summary>
+        /// Specify that a texture resource will be read during the pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify a Texture resource to read during the pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 10]
```diff
@@ -33,6 +49,11 @@ public RenderGraphMutableResource UseDepthBuffer(in RenderGraphMutableResource i
...
+        /// </summary>
+        /// <param name="input">Texture resource that will be read during the pass</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: The Texture resource to read from during the pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 11]
```diff
@@ -42,6 +63,11 @@ public RenderGraphResource ReadTexture(in RenderGraphResource input)
             return input;
         }
 
+        /// <summary>
+        /// Specify that a texture resource will be written to during the pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify a Texture resource to write to during the pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 12]
```diff
@@ -42,6 +63,11 @@ public RenderGraphResource ReadTexture(in RenderGraphResource input)
...
+        /// </summary>
+        /// <param name="input">Texture resource that will be written to during the pass</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: The Texture resource to write to during the pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 13]
```diff
@@ -52,23 +78,42 @@ public RenderGraphMutableResource WriteTexture(in RenderGraphMutableResource inp
             return input;
         }
 
-        public RenderGraphResource UseRendererList(in RenderGraphResource resource)
+        /// <summary>
+        /// Specify that a renderer list resource will be used during this pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify a renderer list resource to read from during the pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 14]
```diff
@@ -52,23 +78,42 @@ public RenderGraphMutableResource WriteTexture(in RenderGraphMutableResource inp
...
+        /// </summary>
+        /// <param name="input">Renderer List resource that will be used during the pass</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: The Renderer List resource to use during the pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 15]
```diff
@@ -52,23 +78,42 @@ public RenderGraphMutableResource WriteTexture(in RenderGraphMutableResource inp
...
+            if (input.type != RenderGraphResourceType.RendererList)
                 throw new ArgumentException("Trying use a resource that is not a renderer list.");
-            m_RenderPass.usedRendererListList.Add(resource);
-            return resource;
+            m_RenderPass.usedRendererListList.Add(input);
+            return input;
         }
+
+        /// <summary>
+        /// Specify the render function used for this pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: Specify the render function to use for this pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs [hunk 16]
```diff
@@ -52,23 +78,42 @@ public RenderGraphMutableResource WriteTexture(in RenderGraphMutableResource inp
...
+        /// A call to this is mandatory for the pass to be valid
+        /// </summary>
+        /// <typeparam name="PassData">Type of the class used to provide data to the Render Pass</typeparam>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphBuilder.cs]: The Type of the class that provides data to the Render Pass

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphObjectPool.cs [hunk 17]
```diff
@@ -3,6 +3,10 @@
 
 namespace UnityEngine.Experimental.Rendering.RenderGraphModule
 {
+    /// <summary>
+    /// Helper class provided in the RenderGraphContext to all Render Passes
+    /// It allows users to do temporary allocations of various objects during a Render Pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphObjectPool.cs]: It allows you to do temporary allocations of various objects during a Render Pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphObjectPool.cs [hunk 18]
```diff
@@ -30,7 +34,13 @@ public void Release(T value)
 
         internal RenderGraphObjectPool() { }
 
-        // APIs for automatically released temporary object allocation
+        /// <summary>
+        /// Allocate a temporary typed array of a specific size
+        /// The array will be released at the end of the Render Pass
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphObjectPool.cs]: Unity releases the array at the end of the Render Pass.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 19]
```diff
@@ -58,6 +64,13 @@ void InitDefaultValues(bool dynamicResolution, bool xrReady)
             }
         }
 
+        /// <summary>
+        /// TextureDesc constructor for a texture using explicit size
+        /// </summary>
+        /// <param name="width">Texture width</param>
+        /// <param name="height">Texture height</param>
+        /// <param name="dynamicResolution">Use dynamic resolution</param>
+        /// <param name="xrReady">Set to true if the texture is used as a render texture in an XR setting</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs]: Set this to true if the Texture is a render texture in an XR setting.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 20]
```diff
@@ -70,6 +83,12 @@ public TextureDesc(int width, int height, bool dynamicResolution = false, bool x
             InitDefaultValues(dynamicResolution, xrReady);
         }
 
+        /// <summary>
+        /// TextureDesc constructor for a texture using a fixed scaling
+        /// </summary>
+        /// <param name="scale">RTHandle scale used for this texture</param>
+        /// <param name="dynamicResolution">Use dynamic resolution</param>
+        /// <param name="xrReady">Set to true if the texture is used as a render texture in an XR setting</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs]: Set this to true if the Texture is a render texture in an XR setting.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 21]
```diff
@@ -82,6 +101,12 @@ public TextureDesc(Vector2 scale, bool dynamicResolution = false, bool xrReady =
             InitDefaultValues(dynamicResolution, xrReady);
         }
 
+        /// <summary>
+        /// TextureDesc constructor for a texture using a functor for scaling
+        /// </summary>
+        /// <param name="func">Function used to determnine the texture size</param>
+        /// <param name="dynamicResolution">Use dynamic resolution</param>
+        /// <param name="xrReady">Set to true if the texture is used as a render texture in an XR setting</param>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs]: Set this to true if the Texture is a render texture in an XR setting.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 22]
```diff
@@ -260,6 +296,11 @@ public int Add(T value)
         List<(int, RTHandle)>               m_AllocatedTextures = new List<(int, RTHandle)>();
 
         #region Public Interface
+        /// <summary>
+        /// Returns the RTHandle associated with the provided resource handle
+        /// </summary>
+        /// <param name="handle">Handle to a texture resource</param>
+        /// <returns>Returns the RTHandle associated with the provided resource handle</returns>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs]: The RTHandle associated with the provided resource handle.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 23]
```diff
@@ -277,6 +318,11 @@ public RTHandle GetTexture(in RenderGraphResource handle)
             return m_TextureResources[handle.handle].rt;
         }
 
+        /// <summary>
+        /// Returns the RendererList associated with the provided resource handle
+        /// </summary>
+        /// <param name="handle">Handle to a Renderer List resource</param>
+        /// <returns>Returns the Renderer List associated with the provided resource handle</returns>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs]: The Renderer List associated with the provided resource handle.

File: com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs [hunk 24]
```diff
@@ -94,11 +119,19 @@ public TextureDesc(ScaleFunc func, bool dynamicResolution = false, bool xrReady
...
+        /// <param name="input"></param>
         public TextureDesc(TextureDesc input)
         {
             this = input;
         }
 
+        /// <summary>
+        /// Hash function
+        /// </summary>
+        /// <returns>Returns the texture descriptor hash</returns>
```
Comment (JordanL8) [inline com.unity.render-pipelines.core/Runtime/RenderGraph/RenderGraphResourceRegistry.cs]: The texture descriptor hash.

**Pull request merged**